8774 search filters

margOnline opened the pull request:
[TP 8774](https://moneyadviceservice.tpondemand.com/restui/board.aspx?#page=board/5580782241563718174&appConfig=eyJhY2lkIjoiRTQ4MDEyNENDOUYxRDU4RDg0RTEyRjNCODBEN0VBMUYifQ==&searchPopup=userstory/8774)

### Summary
This pr:
* builds the frontend view for filtering a search
* parses the filters before sending the request to the cms api

### Approach
* Use the translation yml file to store the values for the search form filters
* create an EvidenceSummarySearchForm object responsible for fleshing out the details of the form
* Use a custom form builder via the Dough form builder

tomas-stefano commented on app/forms/builder.rb:
```diff
@@ -0,0 +1,7 @@
+require 'dough'
+
+class Builder < Dough::Forms::Builder
```
IMHO Builder is a generic name that potentially could cause conflicts. Can we rename to FincapBuilder?

margOnline replied:
😆 I was originally going to use a different name but chose Builder because that is what WPCC uses which you suggested I look at so I thought Builder was a MAS convention. Happy to change.

tomas-stefano commented on app/presenters/evidence_hub_search_form_presenter.rb:
```diff
@@ -0,0 +1,15 @@
+  def class_name
+    object.class.name.underscore
+  end
+end
```
At the moment the form object is only a object with accessors without any methods of handling the form inputs. I really like to hear other's opinions on this, but IMHO this Presenter could be deleted and the filter options could be moved to the Form object, since sounds like Form object responsibility. By the way I didn't understand the class name method **purpose**. Is it important the html class name to be 'evidence_hub_search_form_presenter'?

margOnline replied:
I was thinking that the form object will be responsible for validating the form input. e.g. check that the filter option value is valid, as in one of the values in the yml file. If it is not valid, should we reject it and carry on or return the user to the form. These values will be visible on the url.

The purpose of the `class_name` method prefaces the field parameters with the class name as opposed to class instance e.g.
```
<input class="checkbox" type="checkbox" value="Children (3 - 11)" name="evidence_hub_search_form[client_groups][]" id="evidence_hub_search_form_client_groups_children_3_-_11">
```
vs
```
<input class="checkbox" type="checkbox" value="Children (3 - 11)" name="#<EvidenceHubSearchForm:0x007ff77bcfb7f8>[client_groups][]" id="__EvidenceHubSearchForm:0x007ff77bcfb7f8_client_groups_children_3_-_11">
```

tomas-stefano commented on app/models/filter_option.rb:
```diff
@@ -0,0 +1,11 @@
+  def underscore
+    value.gsub(/[^a-zA-Z. ]/, ' ').strip.split(' ').map(&:downcase).join('_')
+  end
+end
```
Sorry, I didn't understand the purpose of this class. The PR description says "to play nicely with Rails form helpers" and the commit message does not tell much. Can you clarify, please?

margOnline replied:
I can amend the commit message.

margOnline replied:
The collection_check_boxes and collection_radio_buttons act on a collection of objects. It uses methods from the object instance for setting the `value`, `name` and `id` attributes and the content for each checkbox.

willhall88 commented on app/controllers/evidence_hub_controller.rb:
```diff
@@ -1,18 +1,44 @@
+  end
+
+  def filter_params
+    evidence_summary_search_form_params.reject{|k,_| !filter_keys.include?(k) }
```
rather than using a double negative (`reject` and `!`) would it be more readable to use `select` ?

willhall88 commented:
Theres a lot of rubocop offenses on this branch currently. In the process of fixing these could you update the rubocop.yml to point to the remote file please.
e.g. https://github.com/moneyadviceservice/action_plans/blob/master/.rubocop.yml

tomas-stefano commented on app/views/evidence_hub/_checkboxes.html.erb:
```diff
@@ -0,0 +1,10 @@
+    %>
+    <%= option %>
+  <% end %>
+<% end %>
```
Even if the label is wrapped inside a block, the label for and the input name and id are the same, which generates problems with label and input:

```
<label class="checkbox-label" for="evidence_summary_search_form_client_groups">
    <input class="checkbox" name="evidence_summary_search_form[client_groups][]" type="checkbox" value="Children (3 - 11)" id="evidence_summary_search_form_client_groups">
    Children (3 - 11)
</label>
<label class="checkbox-label" for="evidence_summary_search_form_client_groups">
    <input class="checkbox" name="evidence_summary_search_form[client_groups][]" type="checkbox" value="Young people (12 - 16)" id="evidence_summary_search_form_client_groups">
    Young people (12 - 16)
</label>
```

So the label is not working. When you click the last element, the first element is selected. A suggestion would be using an unique identifier (by the index number of the array OR by the parametrize value of the option eg young_people)

tomas-stefano commented on app/views/evidence_hub/_filters.html.erb:
```diff
@@ -0,0 +1,9 @@
+<h4 class="sidepanel__title">Filter your search</h4>
```
I18n

tomas-stefano commented on app/controllers/evidence_hub_controller.rb:
```diff
@@ -2,17 +2,44 @@ class EvidenceHubController < ApplicationController
+
+  def filter_params
+    evidence_summary_search_form_params.select do |k, _|
+      filter_keys.include?(k)
```
I don't understand the purpose of **this** method neither the **filter keys**. Can you clarify, please? Can the FilterParser not just exclude the filters which are empty? That would be more simple.

margOnline replied:
**filter_keys** returns an array of the filter types - client_groups, topics etc. It gets these values from the translation file via the EvidenceSummarySearchForm instance.

**filter_params** returns a subset of the **evidence_summary_search_form params** as only the filter_params need to be formatted to the `blocks` hash required by the cms api. There are no 'empty' filters. If at least one checkbox is not checked for a filter type, that filter type is not passed to the controller by the form

Still to be completed - merging the formatted filter params, or blocks hash which the FilterParser returns, with the other params i.e. document_type and keyword.

tomas-stefano commented on spec/models/evidence_summary_search_form_spec.rb:
```diff
@@ -0,0 +1,51 @@
+      it 'returns true' do
+        search_form.years_of_publication = ['All years', 'Last 2 years']
+        expect(search_form.chosen_options?(
+                 'Last 2 years', :years_of_publication
```
Identation? Did rubocop catch this line?

margOnline replied:
rubocop changed the indentation to this!

willhall88 replied:
how about
```
expect(
  search_form.chosen_options?('Last 2 years', :years_of_publication)
).to be_truthy
```
wdyt?

tomas-stefano replied:
OK you could change to above or leave as it is. Right now I realised that our rubocop needs more tweaking and updating. I will investigate how can improve without causing too much noise and problems.

tomas-stefano commented on spec/models/evidence_summary_search_form_spec.rb:
```diff
@@ -0,0 +1,51 @@
+    context 'when no options have been chosen' do
+      it 'returns false' do
+        expect(search_form.chosen_options?(
+                 'All years', :years_of_publication
```
Identation? Did rubocop catch this line?

margOnline replied:
as above, per the mas standards

tomas-stefano commented on spec/models/evidence_summary_search_form_spec.rb:
```diff
@@ -0,0 +1,51 @@
+    end
+
+    context 'when the second word is plural' do
+      filter = 'client_groups'
```
Can you rename this a to a let, please?

willhall88 commented on app/views/evidence_hub/_filters.html.erb:
```diff
@@ -0,0 +1,9 @@
+<h4 class="sidepanel__title"><%= I18n.t('evidence_hub.filters_header') %></h4>
+
+<h5 class="sidepanel__subtitle">Year of Publication</h5>
```
should this be I18n?

willhall88 commented on config/locales/en.yml:
```diff
@@ -9,9 +9,51 @@ en:
+    filters_header: 'Filter your search'
+    filters:
+      checkbox:
+        - client_groups
```
are these translations? Perhaps I am not understanding, but these strike me as constants that should be within the model?

margOnline commented:
@tomas-stefano @willhall88 I believe I have addressed all your queries and refactored the param parser. Please would you review.

I propose to draw a line under this pr as it is getting pretty big and use this branch as the base for the 2 o/s features, which are listed in the description under TODO.
